sdk/Sdk/FunctionMetadataGenerator.cs

```diff
@@ -368,19 +370,57 @@ private bool TryAddOutputBindingFromMethod(IList<ExpandoObject> bindingMetadata,
 
         private void AddInputTriggerBindingsAndExtensions(IList<ExpandoObject> bindingMetadata, MethodDefinition method)
         {
+            var inputBindingContainsExpression = InputParameterPropertyContainsExpression(method.Parameters);
+
             foreach (ParameterDefinition parameter in method.Parameters)
             {
                 foreach (CustomAttribute parameterAttribute in parameter.CustomAttributes)
                 {
                     if (IsFunctionBindingType(parameterAttribute))
                     {
-                        AddBindingMetadata(bindingMetadata, parameterAttribute, parameter.ParameterType, parameter.Name);
-                        AddExtensionInfo(_extensions, parameterAttribute);
+                        string bindingType = GetBindingType(parameterAttribute);
+                        AddExtensionInfo(_extensions, parameterAttribute, out bool supportsReferenceType);
+
+                        // if this is a trigger and any input binding contains an expression, don't use BindingData
+                        if (Regex.IsMatch(bindingType, "Trigger") && inputBindingContainsExpression)
+                        {
+                            supportsReferenceType = false;
+                        }
+
+                        ExpandoObject binding = BuildBindingMetadataFromAttribute(parameterAttribute, bindingType, parameter.ParameterType, parameter.Name, supportsReferenceType);
+                        bindingMetadata.Add(binding);
                     }
                 }
             }
         }
 
+        private bool InputParameterPropertyContainsExpression(Collection<ParameterDefinition> parameters)
+        {
+            foreach (ParameterDefinition parameter in parameters)
+            {
+                foreach (CustomAttribute parameterAttribute in parameter.CustomAttributes)
```

Comment:
You're missing a check here IsFunctionBindingType which is done in the other loop above. Why do we need to loop through all the parameters twice like this anyways? You could just loop through once, build a list of binding metadatas, then at the end set the deferred property based on whether any used expressions.

Reply:
This was definitly a brute force attempt to get a prototype running so definitly not expecting this to be final form :D I think your suggestion makes sense, you're saying that at the end we can just loop through `bindingMetadata` to check if any input/output bindings use expressions and use that to change the flag for the trigger? This should be a lot more efficient